File: lib/core/models/parking_lots_model.dart
class ParkingLotsModel {
  final String ownerId;
  final String parkingLotName;
  final String postalCode;
  final String prefecture;
  final String city;
  final String addressDetail;
  final String phoneNumber;
  final int capacity;
  final int? availableCapacity;
  final String rentalType;
  final String charge;
  final String? featuresTip;
  final String? nearestStation;
  final String latitude;
  final String longitude;
  final String status;
  final DateTime startDate;
  final DateTime? endDate;
  final DateTime? endStartDatetime;
  final DateTime? endEndDatetime;
  final String? endReason;
  final String? endReasonDetail;
  final String? notes;
  final DateTime? createdDatetime;
  final DateTime? updatedDatetime;

  final ParkingLimitModel? limits;
  final List<ParkingVehicleTypeModel>? vehicleTypes;
  final List<ParkingImageModel>? images;

  ParkingLotsModel({
    required this.ownerId,
    required this.parkingLotName,
    required this.postalCode,
    required this.prefecture,
    required this.city,
    required this.addressDetail,
    required this.phoneNumber,
    required this.capacity,
    this.availableCapacity,
    required this.rentalType,
    required this.charge,
    this.featuresTip,
    this.nearestStation,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.startDate,
    this.endDate,
    this.endStartDatetime,
    this.endEndDatetime,
    this.endReason,
    this.endReasonDetail,
    this.notes,
    this.createdDatetime,
    this.updatedDatetime,
    this.limits,
    this.vehicleTypes,
    this.images,
  });

  factory ParkingLotsModel.fromJson(Map<String, dynamic> json) {
    return ParkingLotsModel(
      ownerId: json['owner_id'] as String,
      parkingLotName: json['parking_lot_name'] as String,
      postalCode: json['postal_code'] as String,
      prefecture: json['prefecture'] as String,
      city: json['city'] as String,
      addressDetail: json['address_detail'] as String,
      phoneNumber: json['phone_number'] as String,
      capacity: json['capacity'] as int,
      availableCapacity: json['available_capacity'] as int?,
      rentalType: json['rental_type'] as String,
      charge: json['charge'] as String,
      featuresTip: json['features_tip'] as String?,
      nearestStation: json['nearest_station'] as String?,
      latitude: json['latitude'].toString(),
      longitude: json['longitude'].toString(),
      status: json['status'] as String,
      startDate: DateTime.parse(json['start_date']),
      endDate: json['end_date'] != null ? DateTime.parse(json['end_date']) : null,
      endStartDatetime: json['end_start_datetime'] != null ? DateTime.parse(json['end_start_datetime']) : null,
      endEndDatetime: json['end_end_datetime'] != null ? DateTime.parse(json['end_end_datetime']) : null,
      endReason: json['end_reason'] as String?,
      endReasonDetail: json['end_reason_detail'] as String?,
      notes: json['notes'] as String?,
      createdDatetime: json['created_datetime'] != null ? DateTime.parse(json['created_datetime']) : null,
      updatedDatetime: json['updated_datetime'] != null ? DateTime.parse(json['updated_datetime']) : null,
      limits: json['limits'] != null ? ParkingLimitModel.fromJson(json['limits']) : null,
      vehicleTypes: json['vehicle_types'] != null
          ? (json['vehicle_types'] as List)
              .map((e) => ParkingVehicleTypeModel.fromJson(e))
              .toList()
          : null,
      images: json['images'] != null
          ? (json['images'] as List)
              .map((e) => ParkingImageModel.fromJson(e))
              .toList()
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    final data = <String, dynamic>{
      'owner_id': ownerId,
      'parking_lot_name': parkingLotName,
      'postal_code': postalCode,
      'prefecture': prefecture,
      'city': city,
      'address_detail': addressDetail,
      'phone_number': phoneNumber,
      'capacity': capacity,
      'available_capacity': availableCapacity,
      'rental_type': rentalType,
      'charge': charge,
      'features_tip': featuresTip,
      'nearest_station': nearestStation,
      'latitude': latitude,
      'longitude': longitude,
      'status': status,
      'start_date': startDate.toIso8601String().split('T')[0], // 这里转换为纯日期字符串
      'end_date': endDate?.toIso8601String().split('T')[0], // 这里转换为纯日期字符串
      'end_start_datetime': endStartDatetime?.toIso8601String(),
      'end_end_datetime': endEndDatetime?.toIso8601String(),
      'end_reason': endReason,
      'end_reason_detail': endReasonDetail,
      'notes': notes,
      'created_datetime': createdDatetime?.toIso8601String(),
      'updated_datetime': updatedDatetime?.toIso8601String(),
    };

    if (limits != null) {
      data['limits'] = limits!.toJson();
    }
    if (vehicleTypes != null) {
      data['vehicle_types'] = vehicleTypes!.map((e) => e.toJson()).toList();
    }
    if (images != null) {
      data['images'] = images!.map((e) => e.toJson()).toList();
    }
    return data;
  }
}

class ParkingLimitModel {
  final int lengthLimit;
  final int widthLimit;
  final int heightLimit;
  final int weightLimit;
  final String carHeightLimit;
  final String tireWidthLimit;
  final String carBottomLimit;

  ParkingLimitModel({
    required this.lengthLimit,
    required this.widthLimit,
    required this.heightLimit,
    required this.weightLimit,
    required this.carHeightLimit,
    required this.tireWidthLimit,
    required this.carBottomLimit,
  });

  factory ParkingLimitModel.fromJson(Map<String, dynamic> json) {
    return ParkingLimitModel(
      lengthLimit: json['length_limit'] as int,
      widthLimit: json['width_limit'] as int,
      heightLimit: json['height_limit'] as int,
      weightLimit: json['weight_limit'] as int,
      carHeightLimit: json['car_height_limit'].toString(),
      tireWidthLimit: json['tire_width_limit'].toString(),
      carBottomLimit: json['car_bottom_limit'].toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'length_limit': lengthLimit,
      'width_limit': widthLimit,
      'height_limit': heightLimit,
      'weight_limit': weightLimit,
      'car_height_limit': carHeightLimit,
      'tire_width_limit': tireWidthLimit,
      'car_bottom_limit': carBottomLimit,
    };
  }
}

class ParkingVehicleTypeModel {
  final String vehicleType;
  final int capacity;

  ParkingVehicleTypeModel({
    required this.vehicleType,
    required this.capacity,
  });

  factory ParkingVehicleTypeModel.fromJson(Map<String, dynamic> json) {
    return ParkingVehicleTypeModel(
      vehicleType: json['vehicle_type'] as String,
      capacity: json['capacity'] as int,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'vehicle_type': vehicleType,
      'capacity': capacity,
    };
  }
}

class ParkingImageModel {
  final String parkingLotId;
  final String imageUrl;

  ParkingImageModel({
    required this.parkingLotId,
    required this.imageUrl,
  });

  factory ParkingImageModel.fromJson(Map<String, dynamic> json) {
    return ParkingImageModel(
      parkingLotId: json['parking_lot_id'] as String,
      imageUrl: json['image_url'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'parking_lot_id': parkingLotId,
      'image_url': imageUrl,
    };
  }
}

class ParkingLotSummaryModel {
  final String parkingLotId;
  final String parkingLotName;

  ParkingLotSummaryModel({
    required this.parkingLotId,
    required this.parkingLotName,
  });

  factory ParkingLotSummaryModel.fromJson(Map<String, dynamic> json) {
    return ParkingLotSummaryModel(
      parkingLotId: json['parking_lot_id'],
      parkingLotName: json['parking_lot_name'],
    );
  }
}


/// 駐車場基本情報のリクエスト
class ParkingLotRequest {
  final String? parking_lot_id;

  ParkingLotRequest({required this.parking_lot_id});

  Map<String, dynamic> toJson() => {'reservation_id': parking_lot_id};
}

File: src/models/parking_lots_model.rs
use chrono::{DateTime, NaiveDate, Utc};
use regex::Regex;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingLotRequest {
    pub owner_id: String,
    pub parking_lot_name: String,
    pub postal_code: String,
    pub prefecture: String,
    pub city: String,
    pub address_detail: String,
    pub phone_number: String,
    pub capacity: i32,
    pub available_capacity: Option<i32>,
    pub rental_type: String,
    pub charge: String,
    pub features_tip: Option<String>,
    pub nearest_station: Option<String>,
    pub latitude: String,
    pub longitude: String,
    pub status: String,
    pub start_date: NaiveDate,
    pub end_date: Option<NaiveDate>,
    pub end_start_datetime: Option<DateTime<Utc>>,
    pub end_end_datetime: Option<DateTime<Utc>>,
    pub end_reason: Option<String>,
    pub end_reason_detail: Option<String>,
    pub notes: Option<String>,
    pub created_datetime: Option<DateTime<Utc>>,
    pub updated_datetime: Option<DateTime<Utc>>,

    pub limits: Option<ParkingLimitRequest>,
    pub vehicle_types: Option<Vec<ParkingVehicleTypeRequest>>,
    pub images: Option<Vec<ParkingImageRequest>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingLimitRequest {
    pub length_limit: Option<i32>,
    pub width_limit: Option<i32>,
    pub height_limit: Option<i32>,
    pub weight_limit: Option<i32>,
    pub car_height_limit: Option<String>,
    pub tire_width_limit: Option<String>,
    pub car_bottom_limit: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingVehicleTypeRequest {
    pub vehicle_type: String,
    pub capacity: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingImageRequest {
    pub parking_lot_id: String,
    pub image_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingLotResponse {
    pub parking_lot_id: String,
    pub owner_id: String,
    pub parking_lot_name: String,
    pub postal_code: String,
    pub prefecture: String,
    pub city: String,
    pub address_detail: String,
    pub phone_number: String,
    pub capacity: i32,
    pub available_capacity: Option<i32>,
    pub rental_type: String,
    pub charge: String,
    pub features_tip: Option<String>,
    pub nearest_station: Option<String>,
    pub latitude: String,
    pub longitude: String,
    pub status: String,
    pub start_date: NaiveDate,
    pub end_date: Option<NaiveDate>,
    pub end_start_datetime: Option<DateTime<Utc>>,
    pub end_end_datetime: Option<DateTime<Utc>>,
    pub end_reason: Option<String>,
    pub end_reason_detail: Option<String>,
    pub notes: Option<String>,
    pub created_datetime: Option<DateTime<Utc>>,
    pub updated_datetime: Option<DateTime<Utc>>,

    pub limits: Option<ParkingLimitModel>,
    pub vehicle_types: Option<Vec<ParkingVehicleTypeModel>>,
    pub images: Option<Vec<ParkingImageModel>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingLimitModel {
    pub parking_lot_id: String,
    pub length_limit: Option<i32>,
    pub width_limit: Option<i32>,
    pub height_limit: Option<i32>,
    pub weight_limit: Option<i32>,
    pub car_height_limit: Option<String>,
    pub tire_width_limit: Option<String>,
    pub car_bottom_limit: Option<String>,
    pub created_datetime: Option<DateTime<Utc>>,
    pub updated_datetime: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingVehicleTypeModel {
    pub id: i32,
    pub parking_lot_id: String,
    pub vehicle_type: String,
    pub capacity: i32,
    pub created_datetime: Option<DateTime<Utc>>,
    pub updated_datetime: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParkingImageModel {
    pub id: i32,
    pub parking_lot_id: String,
    pub image_url: String,
    pub created_datetime: Option<DateTime<Utc>>,
    pub updated_datetime: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, FromRow,Serialize)]
pub struct ParkingLotRow {
    pub parking_lot_id: String,
    pub parking_lot_name: String,
}


impl ParkingLotRequest {
    /// 駐車場登録リクエストの検証処理（バリデーション）
    pub fn validate(&self) -> Result<(), String> {
        // 駐車場名 - VARCHAR(100)
        if self.parking_lot_name.trim().is_empty() {
            return Err("駐車場名は必須です".to_string());
        }
        if self.parking_lot_name.len() > 100 {
            return Err("駐車場名は100文字以内で入力してください".to_string());
        }

        // 郵便番号 - VARCHAR(10)
        if self.postal_code.trim().is_empty() {
            return Err("郵便番号は必須です".to_string());
        }
        if self.postal_code.len() > 10 {
            return Err("郵便番号は10文字以内で入力してください".to_string());
        }
        if !self.is_valid_postal_code(&self.postal_code) {
            return Err("郵便番号は正しい形式（例: 123-4567）で入力してください".to_string());
        }

        // 都道府県 - VARCHAR(50)
        if self.prefecture.trim().is_empty() {
            return Err("都道府県は必須です".to_string());
        }
        if self.prefecture.len() > 50 {
            return Err("都道府県は50文字以内で入力してください".to_string());
        }

        // 市区町村 - VARCHAR(100)
        if self.city.trim().is_empty() {
            return Err("市区町村は必須です".to_string());
        }
        if self.city.len() > 100 {
            return Err("市区町村は100文字以内で入力してください".to_string());
        }

        // 住所詳細 - TEXT (1000 文字制限)
        if self.address_detail.trim().is_empty() {
            return Err("住所詳細は必須です".to_string());
        }
        if self.address_detail.len() > 1000 {
            return Err("住所詳細は1000文字以内で入力してください".to_string());
        }

        // 電話番号 - VARCHAR(50)
        if self.phone_number.trim().is_empty() {
            return Err("電話番号は必須です".to_string());
        }
        if self.phone_number.len() > 50 {
            return Err("電話番号は50文字以内で入力してください".to_string());
        }

        // 台数 - INTEGER
        if self.capacity <= 0 {
            return Err("駐車可能台数は1以上で指定してください".to_string());
        }

        // 金額（charge） - 半角数値、桁数制限等
        if self.charge.trim().is_empty() {
            return Err("料金は必須です".to_string());
        }
        if !self.is_valid_number(&self.charge) {
            return Err("料金は半角数字で入力してください".to_string());
        }

        // 駐車場の特徴（任意） - VARCHAR(500)
        if self.latitude.trim().is_empty() {
            return Err("駐車場の緯度を入力してください".to_string());
        }
        // 駐車場の特徴（任意） - VARCHAR(500)
        if self.longitude.trim().is_empty() {
            return Err("駐車場の経度を入力してください".to_string());
        }

        // その他必要に応じたフィールド追加バリデーション...

        Ok(())
    }

    /// 郵便番号形式（例：123-4567）かどうか判定
    fn is_valid_postal_code(&self, code: &str) -> bool {
        let regex = Regex::new(r"^\d{3}-\d{4}$").unwrap();
        regex.is_match(code)
    }

    /// 半角数字かどうかを検証（料金、制限系等）
    fn is_valid_number(&self, value: &str) -> bool {
        let regex = Regex::new(r"^\d+$").unwrap();
        regex.is_match(value)
    }
}
